content/top-5-programming-language-for-backend-web-programmers.md

```diff
@@ -0,0 +1,1333 @@
+   var updatedAt: Instant? = null
+)
+```
+
+- Repository
+
+```kotlin
+interface PersonRepository : JpaRepository<Person, Long>, JpaSpecificationExecutor<Person> {
+}
+```
+
+- Controller
+
+```kotlin
+@RestController
+class PersonController(
+   private val environment: Environment,
+   private val personRepository : PersonRepository
+) {
+   @GetMapping("/people")
+   fun all() = personRepository.findAll()
+
+   @GetMapping("/people/{id}")
+   fun get(@PathVariable id: Long) = personRepository.findById(id)
+
+   @PostMapping("/people")
+   fun post(@RequestBody person: Person)
+      = ResponseEntity.status(HttpStatus.CREATED).body(personRepository.save(person))
+}
+```
+
+Writing Spring Boot codes using Kotlin is very productive. Many of the reasons are because of:
+
+- **function return inference** which can lead to writing very concise controller functions.
+- **dataclasses**
+  which enables you to create a struct/record data type without **getters** and **setters**. Although you can generate getters and setters through most JVM-based IDEs like **IntelliJ Idea Community** and **Eclipse IDE**, not having to write getters and setters is very intuitive and not time-consuming.
+- **declarative utility functions** such as **map, filter, fold** which Java does not (natively) have, and makes writing loops very, very concise.
+
+```kotlin
+// find something in array
+val found = people.find { p -> p.id == selectedID }
+
+// filter people more than 22 years old
+val ageTwentyTwo = people.filter { p -> p.age > 22 }
+
+// map people's ages
+val onlyAges = people.map { p -> p.age }
+
+```
+
+- Lastly, one of the most important benefit of writing in Kotlin is **null safety**. Rolling-release project model which has very fast-changing RDBMS is very prone to error, and the data that is returned from the RDBMS might not always have a value, which might be a result of **column update**. But the null safety feature in Kotlin prevents your program from getting a random **null pointer exception** which you might will often encounter while writing Java due to Java not having nullable types, so it assumes that every variables are nullable by default, so you can't accidentally unpack a property which is actually **null** while writing Kotlin.
+
+However, as good as Spring + Kotlin is, I would not choose Spring + Kotlin or any JVM based runtime to create a new web-based project these days, because of:
+
+1. **Memory constraints**: JVM apps are notorious for having very big memory usage
+2. **Asynchronous routines not first-class**. Although Kotlin has a neat green threads/coroutines system in the `kotlinx` library, Spring/Spring Boot does not natively support them, and setting them up is painful if you are not used to JVM project management system such as Gradle/Maven
+
+If you or your company already has a Spring Boot project with Java, I really, really recommend migrating to Kotlin because the developer experience and productivity is like night and day.
```

Comment:
```suggestion
If you or your company already has a Spring Boot project with Java, I really, really recommend migrating to Kotlin because the difference in developer experience and productivity is like night and day.
```